### PR TITLE
persist: extract a new shard_source operator from persist_source

### PR DESCRIPTION
**Pull request by danhhz**
Motivations:
- This more clearly reflects the ownership split between persist and
  storage
- This will allow persist to do benchmarking that reflects real world
  usage (all the throughput-sensitive persist usage goes through timely
  dataflows)
- This splits decoding and MFP application into its own timely operator,
  which allows for more precise analysis of future performance
  investigations. (The `Stream<Batch>` output also makes persist look more
  like an arrangement, which is a nice side benefit.)
- This also makes it easier to write unit tests for what has, over time,
  become quite a complex bit of code

### Motivation

   * This PR refactors existing code.

### Tips for reviewer

I _highly_ recommend that you go commit by commit. First one copies in persist_source.rs completely unchanged.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

**Comment by danhhz** [comment]
okay, rebased this! @aljoscha @petrosagg there's some cleanup left, but I wanted to get early feedback on the basic idea

File: src/persist-client/src/operators/shard_source.rs [hunk 1]
```diff
@@ -0,0 +1,510 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! A source that reads from an a persist shard.
```
Comment (aljoscha) [inline src/persist-client/src/operators/shard_source.rs:10]: nit: "an a"

File: src/storage-client/src/source/persist_source.rs [hunk 2]
```diff
@@ -134,465 +126,149 @@ where
...
+    until: Antichain<Timestamp>,
+    mut map_filter_project: Option<&mut MfpPlan>,
+    yield_fn: YFn,
+) -> Stream<G, (Result<Row, DataflowError>, Timestamp, Diff)>
+where
+    G: Scope<Timestamp = mz_repr::Timestamp>,
+    YFn: Fn(Instant, usize) -> bool + 'static,
+{
+    let mut builder =
+        AsyncOperatorBuilder::new(format!("decode_and_mfp({})", name), fetched.scope());
```
Comment (aljoscha) [inline src/storage-client/src/source/persist_source.rs:157]: should be sth like `persist_source::decode_and_mfp`, or `persist_source_decode_and_mfp`

File: src/storage-client/src/source/persist_source.rs [hunk 3]
```diff
@@ -134,465 +126,149 @@ where
...
-        });
-    });
-
-    let mut pinned_stream = Box::pin(async_stream);
-
-    let mut builder_dist = AsyncOperatorBuilder::new(
-        format!("persist_source {}: part distribution", source_id),
-        scope.clone(),
+    let name = source_id.to_string();
+    let (fetched, token) = shard_source(
```
Comment (aljoscha) [inline src/storage-client/src/source/persist_source.rs:130]: don't we eventually have to push the mfp all the way into the shard_source?

**Review by aljoscha — COMMENTED** [review]
did an initial pass, and this seems very good, overall!

File: src/storage-client/src/source/persist_source.rs (hunk 3 above)
Comment (danhhz) [reply src/storage-client/src/source/persist_source.rs:130]: yeah, certainly has the potential to be the most controversial part of this, but I'd like to avoid MapFilterProject being the interface to persist (in the same way we do for Row). TBD a bunch of work for what this actual abstraction looks like, but I'm imagining something morally equivalent to the way Codec abstracts things now (but with necessarily more detail)

File: src/persist-client/src/operators/shard_source.rs (hunk 1 above)
Comment (danhhz) [reply src/persist-client/src/operators/shard_source.rs:10]: done!

File: src/storage-client/src/source/persist_source.rs (hunk 2 above)
Comment (danhhz) [reply src/storage-client/src/source/persist_source.rs:157]: done!

**Review by danhhz — COMMENTED** [review]
okay, I think this is ready to go! I write an initial impl of a non-async decode_and_mfp operator, but 100% got the impl wrong the first time I tried and spooked myself. this change is already scary enough :)

File: src/persist-client/src/operators/shard_source.rs [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+                    );
+                }
+            }
+        }
+    });
+
+    (descs_stream, shutdown_button)
+}
+
+pub(crate) fn shard_source_fetch<K, V, T, D, G>(
```
Comment (petrosagg) [inline src/persist-client/src/operators/shard_source.rs:342]: redundant `T` generic here too

File: src/persist-client/src/operators/shard_source.rs [hunk 5]
```diff
@@ -0,0 +1,467 @@
...
+                                shard_id,
+                                size_in_bytes,
+                                inflight_bytes,
+                                progress,
+                            );
+                        }
+
+                        for part_desc in parts {
+                            // Give the part to a random worker.
+                            let worker_idx = usize::cast_from(Instant::now().hashed()) % num_workers;
```
Comment (petrosagg) [inline src/persist-client/src/operators/shard_source.rs:277]: I think this can be done in a simpler and fairer way (randomness can have hotspots) with this:

```rust
// define this at the top
let worker_round_robin = (0..num_workers).cycle()

let worker_idx = worker_round_robin.next().expect("infinite iterator");
```

File: src/persist-client/src/operators/shard_source.rs [hunk 6]
```diff
@@ -0,0 +1,467 @@
...
+                        cap_set.downgrade(progress.iter());
+                        match progress.elements() {
+                            [ts] => {
+                                current_ts = ts.clone();
+                            }
+                            [] => {
+                                cap_set.downgrade(&[]);
+                                return;
+                            }
+                            _ => panic!("TODO: support partially ordered timestamps in shard_source"),
```
Comment (petrosagg) [inline src/persist-client/src/operators/shard_source.rs:290]: That TODO is pretty hard to fix. DD goes to great lengths to split batches up because timely doesn't support this but it's unclear how we could do the same kind of splitting with persist batches even if timely got the feature.

https://github.com/TimelyDataflow/differential-dataflow/blob/master/src/operators/arrange/arrangement.rs#L572-L575

Given that this isn't going to get fixed any time soon can we put a `TotalOrder` bound until then instead of testing at runtime? With that bound this method becomes available:

```rust
match progress.as_option() {
    Some(ts) => current_ts = ts.clone(),
    None => { .. }
}
```

File: src/persist-client/src/operators/shard_source.rs [hunk 7]
```diff
@@ -0,0 +1,467 @@
...
+                    .return_leased_part(subscription.leased_part_from_exchangeable(leased_part));
+            }
+        });
+    });
+
+    let mut pinned_stream = Box::pin(async_stream);
+
+    let mut builder =
+        AsyncOperatorBuilder::new(format!("shard_source_descs({})", name), scope.clone());
+    let mut flow_control_input = builder.new_input(flow_control_input, Pipeline);
```
Comment (petrosagg) [inline src/persist-client/src/operators/shard_source.rs:243]: Are we disconnecting this in the end? I was convinced by the conversation yesterday that we want this to be disconnected for maximum throughput

File: src/persist-client/src/operators/shard_source.rs [hunk 8]
```diff
@@ -0,0 +1,467 @@
...
+            if let Event::Data(cap, data) = event {
+                // `LeasedBatchPart`es cannot be dropped at this point w/o
+                // panicking, so swap them to an owned version.
+                data.swap(&mut buffer);
+
+                for (_idx, part) in buffer.drain(..) {
+                    let (token, fetched) = fetcher
+                        .fetch_leased_part(fetcher.leased_part_from_exchangeable(part))
+                        .await;
+                    let fetched = fetched.expect("shard_id should match across all workers");
```
Comment (petrosagg) [inline src/persist-client/src/operators/shard_source.rs:396]: We were previously testing the yield function after fetching each part but it looks like now we will fetch an entire slice of times in memory before attempting to yield. Should we move yielding in here?

File: src/storage-client/src/source/persist_source.rs [hunk 9]
```diff
@@ -134,465 +126,156 @@ where
...
+) -> Stream<G, (Result<Row, DataflowError>, Timestamp, Diff)>
+where
+    G: Scope<Timestamp = mz_repr::Timestamp>,
+    YFn: Fn(Instant, usize) -> bool + 'static,
+{
+    // TODO: This operator no longer does anything actually async, and it should
+    // be possible to rewrite it in terms of the normal OperatorBuilder, if we
+    // care. Turns out getting the resumption logic right is pretty tricky
+    // though, and we get a lot of that for "free" right now from the
+    // async/await state machine rust generates for us.
```
Comment (petrosagg) [inline src/storage-client/src/source/persist_source.rs:160]: My :hot_pepper: take is that all operators should be written with `async/await` regardless for exactly this reason!

**Review by petrosagg — COMMENTED** [review]
Overall looks good and seems like a natural split!

File: src/persist-client/src/operators/shard_source.rs (hunk 7 above)
Comment (danhhz) [reply src/persist-client/src/operators/shard_source.rs:243]: done!

File: src/persist-client/src/operators/shard_source.rs (hunk 5 above)
Comment (petrosagg) [reply src/persist-client/src/operators/shard_source.rs:310]: Hm, I'm not sure in what context this suggestion was suggested but this exchange can produce big skews. `Instant::now()` is not a good substitute for a uniform distribution, we're calling in a tight loop and it might produce the same value multiple times in a row.

The round robin solution will always distribute work equally among the workers and it's simple to convince oneself that this is the case.

@frankmcsherry could you type check that the distribution produced by round-robining parts to workers is uniform and that it doesn't suffer from the skew issue that `Instant::now()` modding can fall into?

File: src/storage-client/src/source/persist_source.rs (hunk 9 above)
Comment (petrosagg) [reply src/storage-client/src/source/persist_source.rs]: so previously if we had 10GB worth of parts we would yield before fetching all of them and so if there was a `COUNT(*)` downstream we could process all 10GB of them with much less memory. Maybe I haven't understood the code correctly, but isn't it the case that all 10GB will be loaded here before we can ever try to slice it up in smaller timely batches?

File: src/persist-client/src/operators/shard_source.rs (hunk 5 above)
Comment (frankmcsherry) [reply src/persist-client/src/operators/shard_source.rs:310]: The distribution produced by round robin has the skew issues that round robin might have: if your parts alternate size large, small, then you'll end up only using half of your workers. This might be fine, but a hash is probably better. You don't have to hash the time, you can just hash the round robin index, though.

Comment (petrosagg) [reply src/persist-client/src/operators/shard_source.rs:310]: Ah, but defeating patterns cannot be done with deterministic hashing. The worker assignment produced by (hash(0), hash(1),..) is just as arbitrary and I could construct a malicious input that has a big batch every time that sequence is worker 0. I guess that was the intent of using `Instant::now()` in the first place, which has some amount of randomness.

I tested calling `Instant::now()` in a loop in release mode and it seems to produce different numbers so let's keep it. @dan I would love a comment explaining why we'd like to have some randomness in the selection

File: src/persist-client/src/operators/shard_source.rs (hunk 8 above)
Comment (danhhz) [reply src/persist-client/src/operators/shard_source.rs:428]: > so previously if we had 10GB worth of parts we would yield before fetching all of them and so if there was a COUNT(*) downstream we could process all 10GB of them with much less memory. Maybe I haven't understood the code correctly, but isn't it the case that all 10GB will be loaded here before we can ever try to slice it up in smaller timely batches?

yeah that's how it behaved before, but like I said in my previous response here, I think it's fine? in practice, we won't wait until they're fetched. one part will arrive first, and then this operator will hit an await, yield to timely, and let the next operator run. it's possible that we get a bunch of awaits in a row that all are ready, but unlikely. IMO, it's a better contract for this operator to be allowed to fetch things as fast as it can and do the backpressure in the operator that emits the desc (as the current flow control stuff does)

(now that I type that out, it does strike me that it assumes an even distribution of parts between workers. perhaps we also (instead?) want to plumb the flow_control_input here and give each fetch worker a budget of `flow_control_max_inflight_bytes/scope.peers()`?)

Comment (danhhz) [reply src/persist-client/src/operators/shard_source.rs:428]: mulled this over and realized I could say the above much more simply. proposal: going forward, we backpressure at the granularity of parts using flow_control but "backpressure" within a part with yield_fn

File: src/persist-client/src/operators/shard_source.rs (hunk 5 above)
Comment (danhhz) [reply src/persist-client/src/operators/shard_source.rs:317]: > I would love a comment explaining why we'd like to have some randomness in the selection

done!

File: src/persist-client/src/operators/shard_source.rs [hunk 10]
```diff
@@ -0,0 +1,506 @@
...
+    let mut flow_control_input = builder.new_input_connection(
+        flow_control_input,
+        Pipeline,
+        // Disconnect the flow_control_input from the output capabilities of the
+        // operator. We could leave it connected without risking deadlock so
+        // long as there is a non zero summary on the feedback edge. But it may
+        // be less efficient because the pipeline will be moving in increments
+        // of SUMMARY even though we have potentially dumped a lot more data in
+        // the pipeline because of batch boundaries. Leaving it unconnected
+        // means the pipeline will be able to retire bigger chunks of work.
```
Comment (teskje) [inline src/persist-client/src/operators/shard_source.rs:275]: Excuse the drive-by comment, but I assume you didn't want to include this comment twice :)

Comment (danhhz) [reply src/persist-client/src/operators/shard_source.rs:268]: done!

**Review by petrosagg — APPROVED** [review]
Spoke with Dan on Slack and we think this is good to merge! We can put any improvements in follow up PRs

**Comment by danhhz** [comment]
and aljoscha via slack

> I think it’s good! I like that we have more fine-grained operators, we’ll have more potential for introspection

TFTRs! I'll get this merged ASAP so we can start on followup PRs without having to worry about conflicting with this

**Comment by danhhz** [comment]
Filed https://github.com/MaterializeInc/materialize/issues/16699 for the followups so we don't lose them
